File: xml-performance/test-drive-xml-main-runner/src/main/java/com/steelzack/test/drive/books/XmlTestRunnerWorker.java
package com.steelzack.test.drive.books;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.kohsuke.args4j.CmdLineException;
import org.kohsuke.args4j.CmdLineParser;

import com.steelzack.test.drive.xml.XmlBookDomParserManager;
import com.steelzack.test.drive.xml.XmlBookDomXpathParserManager;
import com.steelzack.test.drive.xml.XmlBookJAXBParserManager;
import com.steelzack.test.drive.xml.XmlBookParserBuilder;
import com.steelzack.test.drive.xml.XmlBookParserManager;
import com.steelzack.test.drive.xml.XmlBookSAXParserManager;

public class XmlTestRunnerWorker {
	private Logger LOG = Logger.getLogger(XmlTestRunner.class.getName());

	private String xmlFile;
	private Integer nIterations;
	private boolean noop;

	public XmlTestRunnerWorker() {
		// CDI
	}

	public XmlTestRunnerWorker(String[] args) throws CmdLineException {
		final XmlTestRunnerOptions options = createNewOptions();
		final CmdLineParser parser = new CmdLineParser(options);
		parser.parseArgument(args);

		xmlFile = options.getBookTestFile();
		nIterations = options.getNumberOfIterations();
		noop = options.isNoop();
	}

	public XmlTestRunnerOptions createNewOptions() {
		return new XmlTestRunnerOptions();
	}

	public boolean runTests() throws Exception {
		callParseIterations(getnIterations(), XmlBookManagerType.DOM);
		callParseIterations(getnIterations(), XmlBookManagerType.DOM_XPATH);
		callParseIterations(getnIterations(), XmlBookManagerType.SAX);
		callParseIterations(getnIterations(), XmlBookManagerType.JAXB);
		return true;
	}

	InputStream getInputStreamFromFile() throws FileNotFoundException {
		return new FileInputStream(new File(xmlFile));
	}

	private void callParseIterations(Integer nIterations, XmlBookManagerType type) throws Exception {
		LOG.log(Level.INFO, "Start of test for {0} with {1} iterations", new Object[] { type, nIterations });
		final Long timeStampNow = System.nanoTime();
		for (int i = 0; i < nIterations; i++) {
			try {
				final InputStream ioStream = getInputStreamFromFile();
				final XmlBookParserManager manager = getManagerFor(ioStream, type);
				manager.init();
			} catch (Exception e) {
				throw new RuntimeException(String.format("Test failed on iteration: %d", i), e);
			}
		}
		LOG.log(Level.INFO, "End of test for {0} with {1} iterations. Took {2} ms",
				new Object[] { type, nIterations, (System.nanoTime() - timeStampNow) / 1000000 });
	}

	private XmlBookParserBuilder getManagerFor(final InputStream ioStream, XmlBookManagerType type) {
		switch (type) {
		case DOM:
			return new XmlBookDomParserManager(ioStream);
		case DOM_XPATH:
			return new XmlBookDomXpathParserManager(ioStream);
		case JAXB:
			return new XmlBookJAXBParserManager(ioStream);
		case SAX:
			return new XmlBookSAXParserManager(ioStream);
		}
		return null;
	}

	public Integer getnIterations() {
		return nIterations;
	}

	public String getXmlFile() {
		return xmlFile;
	}

	public boolean isNoop() {
		return noop;
	}
}

File: xml-performance/test-drive-xml-main-runner/src/main/kotlin/org/jesperancinha/test/drive/books/XmlTestRunnerWorker.kt
package org.jesperancinha.test.drive.books

import org.jesperancinha.test.drive.xml.*
import org.kohsuke.args4j.CmdLineParser
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.InputStream
import java.util.logging.Level
import java.util.logging.Logger

open class XmlTestRunnerWorker {
    private val logger = Logger.getLogger(XmlTestRunner::class.java.name)
    var xmlFile: String? = null
        private set
    private var nIterations: Int? = null
    var isNoop = false
        private set

    constructor()
    constructor(args: Array<String>) {
        val options = createNewOptions()
        val parser = CmdLineParser(options)
        parser.parseArgument(*args)
        xmlFile = options.bookTestFile
        nIterations = options.numberOfIterations
        isNoop = options.isNoop
    }

    fun createNewOptions(): XmlTestRunnerOptions {
        return XmlTestRunnerOptions()
    }

    @Throws(Exception::class)
    fun runTests(): Boolean {
        callParseIterations(getnIterations(), XmlBookManagerType.DOM)
        callParseIterations(getnIterations(), XmlBookManagerType.DOM_XPATH)
        callParseIterations(getnIterations(), XmlBookManagerType.SAX)
        callParseIterations(getnIterations(), XmlBookManagerType.JAXB)
        return true
    }

    @get:Throws(FileNotFoundException::class)
    open val inputStreamFromFile: InputStream
        get() = FileInputStream(File(xmlFile))

    @Throws(Exception::class)
    private fun callParseIterations(nIterations: Int?, type: XmlBookManagerType) {
        logger.log(Level.INFO, "Start of test for {0} with {1} iterations", arrayOf<Any?>(type, nIterations))
        val timeStampNow = System.nanoTime()
        for (i in 0 until nIterations!!) {
            try {
                val ioStream = inputStreamFromFile
                val manager = getManagerFor(ioStream, type)
                manager!!.init()
            } catch (e: Exception) {
                throw RuntimeException(String.format("Test failed on iteration: %d", i), e)
            }
        }
        logger.log(
            Level.INFO,
            "End of test for {0} with {1} iterations. Took {2} ms",
            arrayOf<Any?>(type, nIterations, (System.nanoTime() - timeStampNow) / 1000000)
        )
    }

    private fun getManagerFor(ioStream: InputStream, type: XmlBookManagerType): XmlBookParserBuilder? {
        return when (type) {
            XmlBookManagerType.DOM -> XmlBookDomParserManager(ioStream)
            XmlBookManagerType.DOM_XPATH -> XmlBookDomXpathParserManager(ioStream)
            XmlBookManagerType.JAXB -> XmlBookJAXBParserManager(ioStream)
            XmlBookManagerType.SAX -> XmlBookSAXParserManager(ioStream)
        }
        return null
    }

    open fun getnIterations(): Int? {
        return nIterations
    }
}